Source/WebCore/platform/graphics/ArrayPixelBuffer.h

```diff
@@ -0,0 +1,179 @@
+
+    JSC::ArrayBufferView& data() const LIFETIME_BOUND { return m_data.get(); }
+    Ref<JSC::ArrayBufferView> protectedData() const { return m_data; }
+    Ref<JSC::ArrayBufferView>&& takeData() && { return WTFMove(m_data); }
+    std::span<const uint8_t> span() const LIFETIME_BOUND { return m_data->span(); }
+
+protected:
+    ArrayPixelBuffer(const PixelBufferFormat&, const IntSize&, Ref<JSC::ArrayBufferView>&&);
+
+private:
+    Ref<JSC::ArrayBufferView> m_data;
+};
+
+template <typename JSCTypedArrayView, PixelBuffer::Type pixelBufferType, auto... supportedPixelFormats>
+class TypedArrayPixelBuffer final : public ArrayPixelBuffer {
+public:
+    WEBCORE_EXPORT static Ref<TypedArrayPixelBuffer> create(const PixelBufferFormat&, const IntSize&, JSCTypedArrayView&);
+    WEBCORE_EXPORT static std::optional<Ref<TypedArrayPixelBuffer>> create(const PixelBufferFormat&, const IntSize&, std::span<const uint8_t> data);
+
+    WEBCORE_EXPORT static RefPtr<TypedArrayPixelBuffer> tryCreate(const PixelBufferFormat&, const IntSize&);
+    WEBCORE_EXPORT static RefPtr<TypedArrayPixelBuffer> tryCreate(const PixelBufferFormat&, const IntSize&, Ref<JSC::ArrayBuffer>&&);
+
+    JSCTypedArrayView& data() const LIFETIME_BOUND { return uncheckedDowncast<JSCTypedArrayView>(ArrayPixelBuffer::data()); }
+    Ref<JSCTypedArrayView> protectedData() const { return data(); }
+    Ref<JSCTypedArrayView> takeData() && { return adoptRef(uncheckedDowncast<JSCTypedArrayView>(WTFMove(*this).ArrayPixelBuffer::takeData().leakRef())); }
+
+    Type type() const override { return pixelBufferType; }
+    RefPtr<PixelBuffer> createScratchPixelBuffer(const IntSize&) const override;
+
+private:
+    TypedArrayPixelBuffer(const PixelBufferFormat&, const IntSize&, Ref<JSCTypedArrayView>&&);
+};
+
+template <typename JSCTypedArrayView, PixelBuffer::Type type, auto... supportedPixelFormats>
+Ref<TypedArrayPixelBuffer<JSCTypedArrayView, type, supportedPixelFormats...>> TypedArrayPixelBuffer<JSCTypedArrayView, type, supportedPixelFormats...>::create(const PixelBufferFormat& format, const IntSize& size, JSCTypedArrayView& data)
+{
+    ASSERT(((format.pixelFormat == supportedPixelFormats) || ...));
+    return adoptRef(*new TypedArrayPixelBuffer(format, size, { data }));
+}
+
+template <typename JSCTypedArrayView, PixelBuffer::Type type, auto... supportedPixelFormats>
+std::optional<Ref<TypedArrayPixelBuffer<JSCTypedArrayView, type, supportedPixelFormats...>>> TypedArrayPixelBuffer<JSCTypedArrayView, type, supportedPixelFormats...>::create(const PixelBufferFormat& format, const IntSize& size, std::span<const uint8_t> data)
+{
+    if (!((format.pixelFormat == supportedPixelFormats) || ...)) {
+        ASSERT_NOT_REACHED();
+        return std::nullopt;
+    }
+
+    auto computedBufferSize = PixelBuffer::computeBufferSize(format.pixelFormat, size);
+    if (computedBufferSize.hasOverflowed()) {
+        ASSERT_NOT_REACHED();
+        return std::nullopt;
+    }
```

Comment:
These checks are repeated in the three `create()` functions. Can we make this code shared?